:eyeglasses: sweep docset for areas where `in` wasn't mentioned but should be

BillWagner opened the pull request:
This is the first part of #3949 

In this PR, documents that should have referenced `in` parameters but did not have been updated.

A subsequent PR will update the overload rules for `in` and by value as needed.

BillWagner committed:
sweep docset for areas where `in` wasn't mentioned but should be

rpetrusha commented on docs/csharp/misc/cs1107.md:
````diff
@@ -17,20 +17,22 @@ ms.author: "wiwagn"
 ```  
 // cs1107.cs  
-public static class Test  
+public class Test  
````
A type that contains extension methods has to be `static`, no?

BillWagner replied:
Good catch.

rpetrusha commented on docs/csharp/misc/cs1108.md:
```diff
@@ -17,7 +17,7 @@ ms.author: "wiwagn"
 A parameter cannot have all the specified modifiers; there are too many modifiers on the parameter.  
   
- Certain combinations of parameter modifiers, such as `ref` and `out`, are not allowed because they have mutually exclusive meanings for the compiler.  
+ Certain combinations of parameter modifiers, such as `in`, `ref` and `out`, are not allowed because they have mutually exclusive meanings for the compiler.  
```
comma after `ref`

rpetrusha commented on docs/csharp/misc/cs1611.md:
```diff
@@ -15,9 +15,9 @@ author: "BillWagner"
+The params parameter cannot be declared as in ref or out  
   
- The keywords [ref](../../csharp/language-reference/keywords/ref.md) or [out](../../csharp/language-reference/keywords/out.md) cannot be used with the [params](../../csharp/language-reference/keywords/params.md) keyword.  
+ The keywords [in](../../csharp/language-reference/keywords/in.md), [ref](../../csharp/language-reference/keywords/ref.md) or [out](../../csharp/language-reference/keywords/out.md) cannot be used with the [params](../../csharp/language-reference/keywords/params.md) keyword.  
```
comma after ref

rpetrusha commented on docs/csharp/misc/cs1623.md:
```diff
@@ -15,35 +15,40 @@ author: "BillWagner"
+Iterators cannot have in ref or out parameters  
   
- This error occurs if an iterator method takes a `ref` or `out` parameter. To avoid this error, remove the `ref` or `out` keyword from the method signature.  
+ This error occurs if an iterator method takes an `in`, `ref` or `out` parameter. To avoid this error, remove the `in`, `ref` or `out` keyword from the method signature.  
```
comma after both refs

rpetrusha commented on docs/csharp/misc/cs1628.md:
```diff
@@ -15,9 +15,9 @@ author: "BillWagner"
+Cannot use in ref or out parameter 'parameter' inside an anonymous method, lambda expression, or query expression  
   
- This error occurs if you use a ref or out parameter inside an anonymous method block. To avoid this error, use a local variable or some other construct.  
+ This error occurs if you use an in ref or out parameter inside an anonymous method block. To avoid this error, use a local variable or some other construct.  
```
need some commas here. And fence in, ref, and out?

rpetrusha commented on docs/csharp/misc/cs1951.md:
```diff
@@ -15,13 +15,13 @@ author: "BillWagner"
 ---
 # Compiler Error CS1951
-An expression tree lambda may not contain an out or ref parameter.  
+An expression tree lambda may not contain an in, out or ref parameter.  
```
Are these intended to be literal error messages? If so, it's OK if the punctuation is inconsistent, but otherwise they should be consistent.

BillWagner replied:
I'll be consistent. (I've filed a bug that the compiler only reports ref and out) See dotnet/roslyn#25174

rpetrusha commented on docs/csharp/misc/cs1951.md:
```diff
@@ -15,13 +15,13 @@ author: "BillWagner"
 ## To correct this error  
   
-1.  The only option is to remove the `ref` modifier in the delegate definition and pass in the parameter by value.  
+1.  The only option is to remove the modifier in the delegate definition and pass in the parameter by value.  
```
pass in --> pass (the repetition of in can be slightly confusing)

rpetrusha commented on docs/csharp/programming-guide/classes-and-structs/passing-parameters.md:
```diff
@@ -17,7 +17,7 @@ author: "BillWagner"
 ---
 # Passing Parameters (C# Programming Guide)
-In C#, arguments can be passed to parameters either by value or by reference. Passing by reference enables function members, methods, properties, indexers, operators, and constructors to change the value of the parameters and have that change persist in the calling environment. To pass a parameter by reference, use the `ref` or `out` keyword. For simplicity, only the `ref` keyword is used in the  ...
+In C#, arguments can be passed to parameters either by value or by reference. Passing by reference enables function members, methods, properties, indexers, operators, and constructors to change the value of the parameters and have that change persist in the calling environment. To pass a parameter by reference, use the `in`, `ref` or `out` keyword. For simplicity, only the `ref` keyword is used i ...
```
`in` doesn't conform to the definition about changing and persisting value
Here again, in.md is a disambiguation page without `in` as a parameter modifier.

rpetrusha commented on docs/csharp/programming-guide/classes-and-structs/passing-reference-type-parameters.md:
```diff
@@ -14,7 +14,7 @@ author: "BillWagner"
 ---
 # Passing Reference-Type Parameters (C# Programming Guide)
-A variable of a [reference type](../../../csharp/language-reference/keywords/reference-types.md) does not contain its data directly; it contains a reference to its data. When you pass a reference-type parameter by value, it is possible to change the data pointed to by the reference, such as the value of a class member. However, you cannot change the value of the reference itself; that is, you can ...
+A variable of a [reference type](../../../csharp/language-reference/keywords/reference-types.md) does not contain its data directly; it contains a reference to its data. When you pass a reference-type parameter by value, it is possible to change the data pointed to by the reference, such as the value of a class member. However, you cannot change the value of the reference itself; that is, you can ...
```
A bunch of nits unrelated to your PR:
- "the data pointed to by the reference" becomes confusing, since the reference already is a pointer to the object. I'd interpret this to say that you are changing the value of the reference itself, which is what the next sentence says you cannot do. I think it would be less confusing to say "the data belonging to the referenced object" or some such thing. 
- "that is" is a general statement, but this is a specific example; "for example" would be better here.
- "block" is a method in this context; it would be best to describe it as a method.

rpetrusha commented on docs/csharp/programming-guide/classes-and-structs/passing-reference-type-parameters.md:
```diff
@@ -44,4 +44,6 @@ A variable of a [reference type](../../../csharp/language-reference/keywords/ref
  [Passing Parameters](../../../csharp/programming-guide/classes-and-structs/passing-parameters.md)  
  [Passing Arrays Using ref and out](../../../csharp/programming-guide/arrays/passing-arrays-using-ref-and-out.md)  
  [ref](../../../csharp/language-reference/keywords/ref.md)  
+ [in](../../../csharp/language-reference/keywords/in.md)  
```
here again, the reference to in.md

rpetrusha commented on docs/csharp/programming-guide/classes-and-structs/passing-reference-type-parameters.md:
```diff
@@ -44,4 +44,6 @@ A variable of a [reference type](../../../csharp/language-reference/keywords/ref
  [Passing Arrays Using ref and out](../../../csharp/programming-guide/arrays/passing-arrays-using-ref-and-out.md)  
  [ref](../../../csharp/language-reference/keywords/ref.md)  
+ [in](../../../csharp/language-reference/keywords/in.md)  
+ [out](../../../csharp/language-reference/keywords/out.md)  
```
A better link here is out-parameter-modifier.md.

rpetrusha commented on docs/csharp/programming-guide/classes-and-structs/passing-value-type-parameters.md:
```diff
@@ -14,7 +14,7 @@ author: "BillWagner"
 ---
 # Passing Value-Type Parameters (C# Programming Guide)
-A [value-type](../../../csharp/language-reference/keywords/value-types.md) variable contains its data directly as opposed to a [reference-type](../../../csharp/language-reference/keywords/reference-types.md) variable, which contains a reference to its data. Passing a value-type variable to a method by value means passing a copy of the variable to the method. Any changes to the parameter that take ...
+A [value-type](../../../csharp/language-reference/keywords/value-types.md) variable contains its data directly as opposed to a [reference-type](../../../csharp/language-reference/keywords/reference-types.md) variable, which contains a reference to its data. Passing a value-type variable to a method by value means passing a copy of the variable to the method. Any changes to the parameter that take ...
```
This doesn't describe `in`, though.

rpetrusha commented on docs/csharp/programming-guide/concepts/async/index.md:
```diff
@@ -214,7 +214,7 @@ An async method can also have a `void` return type. This return type is used pri
 An async method that has a `void` return type can’t be awaited, and the caller of a void-returning method can't catch any exceptions that the method throws.  
   
-An async method can't declare [ref](../../../../csharp/language-reference/keywords/ref.md) or [out](../../../../csharp/language-reference/keywords/out.md) parameters, but the method can call methods that have such parameters. Similarly, an async method can't return a value by reference, although it can call methods with ref return values. 
+An async method can't declare [in](../../../../csharp/language-reference/keywords/in.md), [ref](../../../../csharp/language-reference/keywords/ref.md) or [out](../../../../csharp/language-reference/keywords/out.md) parameters, but the method can call methods that have such parameters. Similarly, an async method can't return a value by reference, although it can call methods with ref return values ...
```
comma after ref

rpetrusha commented on docs/csharp/programming-guide/concepts/iterators.md:
```diff
@@ -340,7 +340,7 @@ public class Stack<T> : IEnumerable<T>
  An implicit conversion must exist from the expression type in the `yield return` statement to the return type of the iterator.  
   
- In C#, an iterator method cannot have any `ref` or `out` parameters.  
+ In C#, an iterator method cannot have any `in`, `ref` or `out` parameters.  
```
comma after ref

rpetrusha commented on docs/csharp/programming-guide/statements-expressions-operators/anonymous-methods.md:
```diff
@@ -40,7 +40,7 @@ In versions of C# before 2.0, the only way to declare a [delegate](../../../csha
  A reference to the outer variable `n` is said to be *captured* when the delegate is created. Unlike local variables, the lifetime of a captured variable extends until the delegates that reference the anonymous methods are eligible for garbage collection.  
   
- An anonymous method cannot access the [ref](../../../csharp/language-reference/keywords/ref.md) or [out](../../../csharp/language-reference/keywords/out.md) parameters of an outer scope.  
+ An anonymous method cannot access the [in](../../../csharp/language-reference/keywords/in.md), [ref](../../../csharp/language-reference/keywords/ref.md) or [out](../../../csharp/language-reference/keywords/out.md) parameters of an outer scope.  
```
link to in.md here won't help.

rpetrusha commented on docs/csharp/programming-guide/statements-expressions-operators/lambda-expressions.md:
```diff
@@ -260,7 +260,7 @@ class Test
 -   Variables introduced within a lambda expression are not visible in the outer method.  
   
--   A lambda expression cannot directly capture a `ref` or `out` parameter from an enclosing method.  
+-   A lambda expression cannot directly capture am `in`, `ref` or `out` parameter from an enclosing method.  
```
am --> an
comma after ref

rpetrusha commented on docs/csharp/programming-guide/unsafe-code-pointers/pointer-types.md:
```diff
@@ -109,7 +109,7 @@ Console.ReadLine();
  A pointer can be `null`. Applying the indirection operator to a null pointer causes an implementation-defined behavior.  
   
- Be aware that passing pointers between methods can cause undefined behavior. Examples are returning a pointer to a local variable through an Out or Ref parameter or as the function result. If the pointer was set in a fixed block, the variable to which it points may no longer be fixed.  
+ Be aware that passing pointers between methods can cause undefined behavior. Examples are returning a pointer to a local variable through an In, Out or Ref parameter or as the function result. If the pointer was set in a fixed block, the variable to which it points may no longer be fixed.  
```
unrelated to your PR: are returning --> return
it would be best to lowercase in, out, and ref, and possibly fence them.

rpetrusha commented on docs/csharp/tour-of-csharp/structs.md:
```diff
@@ -34,7 +34,7 @@ With classes, it is possible for two variables to reference the same object and
 If `Point` is a class, the output is 20 because a and b reference the same object. If Point is a struct, the output is 10 because the assignment of `a` to `b` creates a copy of the value, and this copy is unaffected by the subsequent assignment to `a.x`.
 
-The previous example highlights two of the limitations of structs. First, copying an entire struct is typically less efficient than copying an object reference, so assignment and value parameter passing can be more expensive with structs than with reference types. Second, except for `ref` and `out` parameters, it is not possible to create references to structs, which rules out their usage in a nu ...
+The previous example highlights two of the limitations of structs. First, copying an entire struct is typically less efficient than copying an object reference, so assignment and value parameter passing can be more expensive with structs than with reference types. Second, except for `in`, `ref` and `out` parameters, it is not possible to create references to structs, which rules out their usage i ...
```
comma after ref

rpetrusha commented on docs/csharp/language-reference/keywords/method-parameters.md:
```diff
@@ -15,12 +15,14 @@ author: "BillWagner"
+  
 -   [ref](../../../csharp/language-reference/keywords/ref.md)  
   
 -   [out](../../../csharp/language-reference/keywords/out.md)  
```
unrelated to your PR: a better target for `out` is out-parameter-modifier.md

rpetrusha commented on docs/csharp/language-reference/keywords/ref.md:
```diff
@@ -39,7 +39,7 @@ To use a `ref` parameter, both the method definition and the calling method must
 
 An argument that is passed to a `ref` parameter must be initialized before it is passed. This differs from [out](out.md) parameters, whose arguments do not have to be explicitly initialized before they are passed.
```
unrelated to your PR: a better target for `out` is out-parameter-modifier.md

rpetrusha commented on docs/csharp/language-reference/keywords/toc.md:
```diff
@@ -86,6 +86,7 @@
 ### [params](params.md)
+### [in](in.md)
 ### [ref](ref.md)
 ### [out](out.md)
```
unrelated to your PR: this should be removed and out-parameter-modifier.md promoted (unless you consider a generic type parameter in a method signature a method parameter, in which case out-generic-modifier.md should be added.

BillWagner replied:
I added the other uses for out and in to the TOC.

rpetrusha commented on docs/csharp/misc/cs1601.md:
```diff
@@ -17,22 +17,27 @@ ms.author: "wiwagn"
 Method or delegate parameter cannot be of type 'type'  
   
- Some types in the .NET Framework class library, for example, <xref:System.TypedReference>, <xref:System.RuntimeArgumentHandle> and <xref:System.ArgIterator> cannot be used as [ref](../../csharp/language-reference/keywords/ref.md) or [out](../../csharp/language-reference/keywords/out.md) parameters because they could potentially be used to perform unsafe operations.  
+ Some types in the .NET Framework class library, for example, <xref:System.TypedReference>, <xref:System.RuntimeArgumentHandle> and <xref:System.ArgIterator> cannot be used as [in](../../csharp/language-reference/keywords/in.md), [ref](../../csharp/language-reference/keywords/ref.md) or [out](../../csharp/language-reference/keywords/out.md) parameters because they could potentially be used to per ...
```
unrelated to your PR: a better target for `out` is out-parameter-modifier.md

rpetrusha reviewed (changes requested):
I've left a number of comments and suggestions, @BillWagner. 
One nit unrelated to your PR that you might want to address is the numerous references to out.md, which is a disambiguation topic. These are probably best changed to out-parameter-modifier.md.
But the major issue is that there are numerous links to the in keyword, which again is a disambiguation page that doesn't disambiguate the `in` parameter modifier, and there is no language keyword that documents the `in` parameter modifier.

BillWagner commented on docs/csharp/language-reference/keywords/async.md:
```diff
@@ -73,7 +73,7 @@ An async method can have the following return types:
 - Starting with C# 7, any type that has an accessible `GetAwaiter` method. The `System.Threading.Tasks.ValueTask<TResult>` type is one such implementation. It is available by adding the NuGet package `System.Threading.Tasks.Extensions`. 
 
-The async method can't declare any [ref](../../../csharp/language-reference/keywords/ref.md) or [out](../../../csharp/language-reference/keywords/out.md) parameters, nor can it have a <!-- [reference return value](../../programming-guide/classes-and-structs/ref-returns.md) -->reference return value, but it can call methods that have such parameters.  
+The async method can't declare any [in](../../../csharp/language-reference/keywords/in.md) [ref](../../../csharp/language-reference/keywords/ref.md) or [out](../../../csharp/language-reference/keywords/out.md) parameters, nor can it have a <!-- [reference return value](../../programming-guide/classes-and-structs/ref-returns.md) -->reference return value, but it can call methods that have such par ...
```
I don't know why the reference return was commented out. I put it back.
As long as I'm changing so many files, I did update the references from out.md to out-parameter-modifier.md throughout the docs where that was the better choice.

BillWagner commented on docs/csharp/language-reference/keywords/in-generic-modifier.md:
```diff
@@ -18,7 +18,7 @@ For generic type parameters, the `in` keyword specifies that the type parameter
  Contravariance enables you to use a less derived type than that specified by the generic parameter. This allows for implicit conversion of classes that implement variant interfaces and implicit conversion of delegate types. Covariance and contravariance in generic type parameters are supported for reference types, but they are not supported for value types.  
   
- A type can be declared contravariant in a generic interface or delegate if it is used only as a type of method arguments and not used as a method return type. `Ref` and `out` parameters cannot be variant.  
+ A type can be declared contravariant in a generic interface or delegate if it is used only as a type of method arguments and not used as a method return type. `In`, `ref` and `out` parameters cannot be variant.  
```
You may have found a C# language spec issue.  Historically, `out` and `ref` args must be invariant. (Which is a more clear way of saying they cannot be variant, which means they can't be covariant or contravariant.) However, it would seem that `in` arguments could be contravariant. It might not be super useful, because `in` arguments are typically structs.

BillWagner replied:
I reworded. I now say that in ref and out parameters must be invariant.

BillWagner commented on docs/csharp/language-reference/keywords/method-parameters.md:
```diff
@@ -15,12 +15,14 @@ author: "BillWagner"
 ---
 # Method Parameters (C# Reference)
-If a parameter is declared for a method without [ref](../../../csharp/language-reference/keywords/ref.md) or [out](../../../csharp/language-reference/keywords/out.md), the parameter can have a value associated with it. That value can be changed in the method, but the changed value will not be retained when control passes back to the calling procedure. By using a method parameter keyword, you can  ...
+If a parameter is declared for a method without [in](../../../csharp/language-reference/keywords/in.md), [ref](../../../csharp/language-reference/keywords/ref.md) or [out](../../../csharp/language-reference/keywords/out.md), the parameter can have a value associated with it. That value can be changed in the method, but the changed value will not be retained when control passes back to the calling ...
```
I put the WIP designation back on this PR (I'll add a new file for in-parameter-modifier.md)
I've updated all the references to out-parameter-modifier.md where that is a better match.

BillWagner commented on docs/csharp/language-reference/keywords/out-parameter-modifier.md:
```diff
@@ -35,15 +35,15 @@ class CS0663_Example
  For information about passing arrays, see [Passing Arrays Using ref and out](../../../csharp/programming-guide/arrays/passing-arrays-using-ref-and-out.md).  
   
- You can't use the `ref` and `out` keywords for the following kinds of methods:  
+ You can't use the `in`, `ref` and `out` keywords for the following kinds of methods:  
```
I did a global search, and I think I fixed all of them.

BillWagner commented on docs/csharp/language-reference/keywords/toc.md:
```diff
@@ -86,6 +86,7 @@
 ### [lock Statement](lock-statement.md)
 ## [Method Parameters](method-parameters.md)
 ### [params](params.md)
+### [in](in.md)
```
placeholder is there. I updated all links as appropriate. (Content coming on moday)

BillWagner committed:
respond to first round of feedback

pkulikov commented on docs/csharp/language-reference/keywords/in-generic-modifier.md:
```diff
@@ -18,7 +18,7 @@ For generic type parameters, the `in` keyword specifies that the type parameter
  Contravariance enables you to use a less derived type than that specified by the generic parameter. This allows for implicit conversion of classes that implement variant interfaces and implicit conversion of delegate types. Covariance and contravariance in generic type parameters are supported for reference types, but they are not supported for value types.  
   
- A type can be declared contravariant in a generic interface or delegate if it is used only as a type of method arguments and not used as a method return type. `In`, `ref` and `out` parameters cannot be variant.  
+ A type can be declared contravariant in a generic interface or delegate if it is used only as a type of method arguments and not used as a method return type. `In`, `ref`, and `out` parameters must be invariant (meaning they are never covariant or contravariant.)  
```
*never* -> *neither* in *...(meaning they never covariant or...*
Plus the final point should be out of parentheses?

BillWagner committed:
create in-modifer article

Also, update samples and snippets and make corresponding changs to 'ref' and 'out' article.

BillWagner committed:
fix build errors

BillWagner committed:
fix one more build error

BillWagner committed:
wording change based on feedback.

BillWagner commented:
@rpetrusha I've updated everything from the previous review.

I've also added the new page for the in parameter modifier.

rpetrusha commented on docs/csharp/programming-guide/classes-and-structs/passing-reference-type-parameters.md:
```diff
@@ -14,7 +14,7 @@ author: "BillWagner"
 ---
 # Passing Reference-Type Parameters (C# Programming Guide)
-A variable of a [reference type](../../../csharp/language-reference/keywords/reference-types.md) does not contain its data directly; it contains a reference to its data. When you pass a reference-type parameter by value, it is possible to change the data pointed to by the reference, such as the value of a class member. However, you cannot change the value of the reference itself; that is, you can ...
+A variable of a [reference type](../../../csharp/language-reference/keywords/reference-types.md) does not contain its data directly; it contains a reference to its data. When you pass a reference-type parameter by value, it is possible to change the data belonging to the referenced object, such as the value of a class member. However, you cannot change the value of the reference itself; for examp ...
```
in doesn't really apply here

rpetrusha commented on docs/csharp/programming-guide/unsafe-code-pointers/pointer-types.md:
```diff
@@ -109,7 +109,7 @@ Console.ReadLine();
  A pointer can be `null`. Applying the indirection operator to a null pointer causes an implementation-defined behavior.  
   
- Be aware that passing pointers between methods can cause undefined behavior. Examples are returning a pointer to a local variable through an Out or Ref parameter or as the function result. If the pointer was set in a fixed block, the variable to which it points may no longer be fixed.  
+ Be aware that passing pointers between methods can cause undefined behavior. Examples are return a pointer to a local variable through an `in`, `out` or `ref` parameter or as the function result. If the pointer was set in a fixed block, the variable to which it points may no longer be fixed.  
```
the original "returning" is preferable, though the sentence could use some rewriting, such as, "Consider a method that returns a pointer..."

rpetrusha reviewed (changes requested):
You've done an impressive job with this revision, @BillWagner. I've left a number of comments, but this is getting close to being ready to merge.

BillWagner commented on docs/csharp/language-reference/keywords/out-parameter-modifier.md:
````diff
@@ -35,15 +33,16 @@ class CS0663_Example
 ```
   
-Overloading is legal, however, if one method takes a `ref` or `out` argument and the other uses neither, like this:  
+Overloading is legal, however, if one method takes a `ref`, `in`, or `out` argument and the other has none of those modifiers, like this:  
````
I felt bad too. I added a sentence now. It's not perfect, and I'll add more in the next PR, but it helps.

BillWagner commented on docs/csharp/language-reference/keywords/ref.md:
```diff
@@ -35,35 +33,41 @@ When used in a method's parameter list, the `ref` keyword indicates that an argu
+[!code-csharp-interactive[csrefKeywordsMethodParams#6](../../../../samples/snippets/csharp/language-reference/keywords/in-ref-out-modifier/RefParameterModifier.cs#1)]
 
-An argument that is passed to a `ref` parameter must be initialized before it is passed. This differs from [out](out.md) parameters, whose arguments do not have to be explicitly initialized before they are passed.
+An argument that is passed to a `ref` parameter must be initialized before it is passed. This differs from [out](out-parameter-modifier.md) parameters, whose arguments do not have to be explicitly initialized before they are passed.
```
I wasn't sure, but probably. I added it.

BillWagner commented on docs/csharp/methods.md:
```diff
@@ -113,7 +113,7 @@ The following example defines a class (which is a reference type) named `SampleR
 ### Passing parameters by reference ###
 
-You pass a parameter by reference when you want to change the value of an argument in a method and want to refect that change when control returns to the calling method. To pass a parameter by reference, you use the `ref` or `out` keyword.
+You pass a parameter by reference when you want to change the value of an argument in a method and want to refect that change when control returns to the calling method. To pass a parameter by reference, you use the `ref` or `out` keyword. You can also pass a value by reference to avoid copying using the `in` keyword.
```
updated for out and ref as well

BillWagner committed:
respond to edits

BillWagner committed:
commit changes from that unsaved window

Yeah, I did that. Didn't save before committing.

rpetrusha commented:
This looks really good, @BillWagner. It's ready to merge when you want.
